src/Microsoft.TestPlatform.CommunicationUtilities/SocketClient.cs

```diff
@@ -0,0 +1,102 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace Microsoft.VisualStudio.TestPlatform.CommunicationUtilities
+{
+    using System;
+    using System.IO;
+    using System.Net;
+    using System.Net.Sockets;
+    using System.Threading;
+    using System.Threading.Tasks;
+    using Microsoft.VisualStudio.TestPlatform.CommunicationUtilities.Interfaces;
+    using Microsoft.VisualStudio.TestPlatform.ObjectModel;
+
+    /// <summary>
+    /// Communication client implementation over sockets.
+    /// </summary>
+    public class SocketClient : ICommunicationClient
+    {
+        private readonly CancellationTokenSource cancellation;
+        private readonly TcpClient tcpClient;
+        private readonly Func<Stream, ICommunicationChannel> channelFactory;
+        private ICommunicationChannel channel;
+        private bool stopped;
+
+        public SocketClient()
+            : this(stream => new LengthPrefixCommunicationChannel(stream))
+        {
+        }
+
+        protected SocketClient(Func<Stream, ICommunicationChannel> channelFactory)
+        {
+            // Used to cancel the message loop
+            this.cancellation = new CancellationTokenSource();
+            this.stopped = false;
+
+            this.tcpClient = new TcpClient();
+            this.channelFactory = channelFactory;
+        }
+
+        /// <inheritdoc />
+        public event EventHandler<ConnectedEventArgs> ServerConnected;
+
+        /// <inheritdoc />
+        public event EventHandler<DisconnectedEventArgs> ServerDisconnected;
+
+        /// <inheritdoc />
+        public void Start(string connectionInfo)
+        {
+            this.tcpClient.ConnectAsync(IPAddress.Loopback, int.Parse(connectionInfo))
+                .ContinueWith(this.OnServerConnected);
+        }
+
+        /// <inheritdoc />
+        public void Stop()
+        {
+            if (!this.stopped)
+            {
+                EqtTrace.Info("SocketClient: Stop: Cancellation requested. Stopping message loop.");
+                this.cancellation.Cancel();
+            }
+        }
+
+        private void OnServerConnected(Task connectAsyncTask)
+        {
+            if (connectAsyncTask.IsFaulted)
+            {
+                // Throw an exception
+            }
+
+            this.channel = this.channelFactory(this.tcpClient.GetStream());
+            if (this.ServerConnected != null)
+            {
+                this.ServerConnected.Invoke(this, new ConnectedEventArgs(this.channel));
```

Comment:
Suggestion : Use [SafeInvoke](https://github.com/Microsoft/vstest/blob/c701070abb126853b58800cedd7554482471ac7d/src/Microsoft.TestPlatform.CoreUtilities/Utilities/MulticastDelegateUtilities.cs#L26) as it takes care of EqtTrace logging as well.